File: 295-find-median-from-data-stream/find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.small = []
        self.large = []

    def addNum(self, num: int) -> None:
        heapq.heappush(self.small, -1 * num)

        if (self.small and self.large and -1 * self.small[0] > self.large[0] ):
            heapq.heappush(self.large, -1 * heapq.heappop(self.small))

        if len(self.small) - 1 > len(self.large):
            heapq.heappush(self.large, -1 * heapq.heappop(self.small))
        
        if len(self.small) < len(self.large) - 1:
            heapq.heappush(self.small, -1 * heapq.heappop(self.large))

    def findMedian(self) -> float:
        if len(self.small) > len(self.large): return -1 * self.small[0]
        if len(self.small) < len(self.large): return self.large[0]
        return (-1 * self.small[0] + self.large[0]) / 2.0
        


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 295-find-median-from-data-stream/find-median-from-data-stream.cpp
class MedianFinder {
    priority_queue<int> small;
    priority_queue<int, vector<int>, greater<int>> large;

public:
    MedianFinder() {

    }
    
    void addNum(int num) {
        small.push(num);

        if (small.size()>0 && large.size()>0 && small.top() > large.top()){
            large.push(small.top());
            small.pop();
        }

        if (small.size() > large.size() + 1){
            large.push(small.top());
            small.pop(); 
        }

        if (small.size() + 1 < large.size()){
            small.push(large.top());
            large.pop();
        }
    }
    
    double findMedian() {
        if (small.size() > large.size()) return small.top();
        if (small.size() < large.size()) return large.top();
        return  (small.top() + large.top())/2.0;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */